File: 2025/day10/main.cpp
#include <bits/stdc++.h>
#include <cctype>
#include <queue>
#include <sstream>
#include <unordered_set>
#include <vector>
using namespace std;

struct State {
  vector<bool> current;
  int presses;
};

int solve(vector<bool> &target, vector<vector<int>> &buttons) {
  vector<bool> start(target.size(),
                     false); // initially all the switches will be off
  queue<State> q;
  q.push({start, 0});
  unordered_set<vector<bool>> visited;

  while (!q.empty()) {
    State cur = q.front();
    q.pop();
    visited.insert(cur.current);
    if (cur.current == target) {
      return cur.presses;
    }

    for (auto button : buttons) {
      vector<bool> next = cur.current;
      for (auto i : button) {
        // simply toggle the state;
        next[i] = !next[i];
      }

      if (!visited.count(next)) {
        q.push({next, cur.presses + 1});
        visited.insert(next);
      }
    }
  }
  return 0;
}

int main() {
  const string filename = "inputs.txt";

  ifstream file(filename);
  if (!file.is_open()) {
    cerr << "Error Unable to open the file\n";
    return 0;
  }

  string line;
  int total = 0;

  while (getline(file, line)) {
    vector<bool> target; // target holds the boolean of the final switches
    vector<vector<int>> buttons; // the buttons consists of an array of
                                 // acceptable buttons we can press

    stringstream ss(line);
    char ch; // helps us to go char by char and put values into the array

    while (ss >> ch) {
      // take the starting occurance of the [ and go on until u find ] and push
      // them all in to the array once
      if (ch == '[') {
        while (ss >> ch && ch != ']') {
          ch == '#' ? target.push_back(true) : target.push_back(false);
        }
      }
      // handling the buttons
      else if (ch == '(') {
        vector<int> currGroup;
        int num;
        while (ss >> ch) {
          if (isdigit(ch)) {
            ss.putback(ch);
            ss >> num;
            currGroup.push_back(num);
          } else if (ch == ')') {
            break;
          }
        }
        buttons.push_back(currGroup);
      } else if (ch == '{') {
        break;
      }
    }
    auto print = [&]() -> void {
      cout << "Target: [ ";
      for (auto t : target)
        cout << t << " ";
      cout << " ]\n";

      cout << " Buttons: [ ";
      for (auto &v : buttons) {
        cout << "( ";
        for (auto x : v)
          cout << x << ",";
        cout << " )";
      }
      cout << " ]\n";
    };
    int ans = solve(target, buttons);
    total += ans;
  }
  cout << total << "\n";
  return 0;
}

File: 2025/day10/src/main.rs
/* 0110 -- target
 * 0001
 * 0101
 * 0010
 * 0011
 * 1010
 * 1100
 * Think of toggling the bit what we use is OR but we should toggle things even when they
 * are the same means one thing which is XOR thats the main trick
 * assuming the given switches as binary positions and we need to find the things
 * that produce the target when XORed and they should be in less steps
 */

/*
 * But think of even more cooler way is to use BFS
 */

use std::{
    collections::{HashSet, VecDeque},
    fs,
};

fn main() {
    let input = fs::read_to_string("../test.txt").unwrap();
    let mut total = 0;

    for line in input.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        // First part is the target
        let target_str = parts[0];
        let target: Vec<bool> = target_str
            .trim_matches(['[', ']'])
            .chars()
            .map(|c| c == '#')
            .collect();
        // println!("{:?}", target)
        // basically converts # to true and . to false
        let mut buttons = Vec::new();
        for part in &parts[1..] {
            if part.starts_with('{') {
                break; // ignore the last {} for part 1
            }

            let nums = part
                .trim_matches(['(', ')'])
                .split(',')
                .map(|x| x.parse::<usize>().unwrap())
                .collect::<Vec<_>>();
            buttons.push(nums);
        }
        // println!("{:?}", buttons); get the values as an array
        let anwser = solve(&target, &buttons);
        println!("{} -> {}", line, anwser);
        total += anwser;
    }
    println!("Total: {}", total)
}

fn solve(target: &[bool], buttons: &[Vec<usize>]) -> usize {
    let start = vec![false; target.len()];
    println!("{:?}", start);
    let mut queue = VecDeque::new();
    queue.push_back((start.clone(), 0)); // current state and number of presses

    let mut visited = HashSet::new();
    visited.insert(start);

    while let Some((state, presses)) = queue.pop_front() {
        if state == target {
            return presses;
        }

        for button in buttons {
            let mut next = state.clone();
            for &index in button {
                next[index] = !next[index];
            }

            if visited.insert(next.clone()) {
                queue.push_back((next, presses + 1));
            }
        }
    }
    unreachable!("Target cannot be reached");
}
